core/src/main/scala/spire/optional/Perm.scala

```diff
@@ -0,0 +1,79 @@
+package spire
+package optional
+
+import scala.collection.compat._
+import spire.scalacompat.SeqLike
+
+import cats.kernel.Eq
+import spire.algebra.{Action, Group}
+import spire.algebra.partial.PartialAction
+import spire.syntax.cfor._
+import spire.syntax.group._
+import spire.util._
+
+/**
+ * Represents a permutation encoded as a map from preimages to images, including
+ * only pairs that are moved by the permutation (so the identity is Map.empty).
+ */
+class Perm private(val mapping: Map[Int, Int]) extends (Int => Int) {
+  assert(mapping.forall(Function.tupled(_ != _)), "Perm contains a redundant mapping.")
```

Comment:
this might now be redundant with `Perm.apply` filter below

Reply:
Good catch; this isn't needed anymore.
Perhaps I should turn this into a property test? (The risk is that this causes `===` to fail, so I could test equality of identity mappings rather than the specific implementation detail).